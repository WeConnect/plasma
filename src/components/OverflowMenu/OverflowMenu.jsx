import cx from 'classnames';
import { map } from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { getDataAttrs, getDataProps } from '../../dataUtils';
import OverflowMenuItem from './OverflowMenuItem.jsx';
import style from './style.scss';

const direction = { RIGHT: 'right', LEFT: 'left' };

class OverflowMenu extends React.Component {
  constructor() {
    super();
    this.state = {
      revealed: false,
    };
  }

  handleMouseEnter = () => {
    if (!this.props.disabled) {
      this.setState({ revealed: true });
    }
  };

  handleMouseLeave = () => {
    this.setState({ revealed: false });
  };

  handleClick = event => {
    this.setState({ revealed: false });
    this.props.onClick(event);
  };

  renderLabel = () => {
    const { title } = this.props;
    if (title) {
      return <div className={style.labelWithText}>{title} &#9662;</div>;
    }

    return <div className={style.overflow} />;
  };

  render() {
    const revealableStyle = cx(style.revealable, {
      [style.revealed]: this.state.revealed,
    });
    const revealableListStyle = cx(style.revealableList, {
      [style.openLeft]: this.props.openDirection && this.props.openDirection === direction.LEFT,
    });
    return (
      <div
        {...getDataAttrs(this.props.data)}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
        className={style.container}
      >
        <div
          ref={c => {
            this.title = c;
          }}
          className={style.action}
        >
          {this.renderLabel()}
        </div>
        <div
          ref={c => {
            this.revealable = c;
          }}
          className={revealableStyle}
        >
          <div className={style.revealableTopWrapper}>
            <div className={style.revealableTop}>{this.renderLabel()}</div>
          </div>
          <ol className={revealableListStyle}>
            {map(this.props.options, option => {
              return (
                <OverflowMenuItem
                  key={option.key}
                  optionKey={option.key}
                  text={option.text}
                  onClick={this.handleClick}
                />
              );
            })}
          </ol>
        </div>
      </div>
    );
  }
}

OverflowMenu.defaultProps = {
  options: {},
  openDirection: direction.RIGHT,
};

OverflowMenu.propTypes = {
  disabled: PropTypes.bool,
  options: PropTypes.array,
  onClick: PropTypes.func,
  openDirection: PropTypes.oneOf([direction.LEFT, direction.RIGHT]),
  ...getDataProps(),
};

OverflowMenu.displayName = 'Plasma@OverflowMenu';

export default OverflowMenu;
