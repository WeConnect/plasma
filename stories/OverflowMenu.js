import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { linkTo } from '@storybook/addon-links';
import OverflowMenu from '../src/components/OverflowMenu/OverflowMenu';

const options = [
  { key: 'bury', text: 'Bury' },
  { key: 'marry', text: 'Marry' },
  { key: 'screw', text: 'Screw' },
];

storiesOf('OverflowMenu', module)
  .add('default', () => <OverflowMenu options={options} onClick={action('clicked')} />)
  .add('open left', () => (
    <div style={{ float: 'right' }}>
      <OverflowMenu openDirection="left" options={options} onClick={action('clicked')} />
    </div>
  ));
