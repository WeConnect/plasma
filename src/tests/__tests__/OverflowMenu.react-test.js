/* eslint-disable */

import React from 'react';
import OverflowMenu from '../../components/OverflowMenu/OverflowMenu';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

describe('OverflowMenu Component', () => {
  test('render', () => {
    const component = renderer.create(
      <OverflowMenu />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});