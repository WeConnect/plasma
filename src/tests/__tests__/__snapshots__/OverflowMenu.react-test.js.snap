// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverflowMenu Component render 1`] = `
<div
  className="container"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="action"
  >
    <div
      className="overflow"
    />
  </div>
  <div
    className="revealable"
  >
    <div
      className="revealableTopWrapper"
    >
      <div
        className="revealableTop"
      >
        <div
          className="overflow"
        />
      </div>
    </div>
    <ol
      className="revealableList"
    />
  </div>
</div>
`;
